File: 3563.lexicographically-smallest-string-after-adjacent-removals.cpp
#
# @lc app=leetcode id=3563 lang=cpp
#
# [3563] Lexicographically Smallest String After Adjacent Removals
#
# @lc code=start
class Solution {
public:
    unordered_map<string, string> memo;
    
    bool isConsecutive(char a, char b) {
        int diff = abs(a - b);
        return diff == 1 || diff == 25; // 25 for 'a' and 'z'
    }
    
    string lexicographicallySmallestString(string s) {
        if (memo.count(s)) return memo[s];
        
        string result = s;
        
        // Try all possible removals
        for (int i = 0; i < (int)s.length() - 1; i++) {
            if (isConsecutive(s[i], s[i+1])) {
                string next = s.substr(0, i) + s.substr(i + 2);
                string candidate = lexicographicallySmallestString(next);
                if (candidate < result) {
                    result = candidate;
                }
            }
        }
        
        memo[s] = result;
        return result;
    }
};
# @lc code=end

File: 3563.lexicographically-smallest-string-after-adjacent-removals.java
#
# @lc app=leetcode id=3563 lang=java
#
# [3563] Lexicographically Smallest String After Adjacent Removals
#
# @lc code=start
class Solution {
    public String lexicographicallySmallestString(String s) {
        // Use recursion with memoization to explore all possible removal sequences
        return dfs(s, new HashMap<>());
    }
    
    private String dfs(String s, Map<String, String> memo) {
        if (memo.containsKey(s)) return memo.get(s);
        String minStr = s;
        for (int i = 0; i < s.length() - 1; i++) {
            char a = s.charAt(i), b = s.charAt(i+1);
            if (isConsecutive(a, b)) {
                String next = s.substring(0, i) + s.substring(i+2);
                String candidate = dfs(next, memo);
                if (candidate.compareTo(minStr) < 0) {
                    minStr = candidate;
                }
            }
        }
        memo.put(s, minStr);
        return minStr;
    }
    
    private boolean isConsecutive(char a, char b) {
        int diff = Math.abs(a - b);
        return diff == 1 || diff == 25; // 25 for circular adjacency (a-z)
    }
}
# @lc code=end